### ISSUE TITLE
Question: meaning of cache-friendly

### ISSUE THREAD

**Issue by photoszzt** [microsoft/snmalloc#256]
I'm using snmalloc in a rust project. I find that the cache-friendly option exposed in the crate helps. I wonder what it does. Is there any documentation?

**Comment by davidchisnall** [comment]
This can probably be removed, because I think it only made a difference in a contrived microbenchmark from the supermalloc project, which didn't touch any of the memory between alloc and free, so was very sensitive to metadata placement.

By default, snmalloc uses the first word of free'd objects to store metadata.  For some power-of-two-sized allocations, this can mean that the metadata ends up in the same associativity set, increasing cache pressure.  This option tries to put the metadata at a different offset to try to avoid the problem.

We have not found a real workload where it makes a positive difference.  

As I recall, we kept the code because the extension point that it uses is one that might be useful for pointer signing, CHERI, and other mitigations, though I'm not completely sure that it would be.  @mjp41, what do you think?  Should we just remove that code?

**Comment by mjp41** [comment]
> We have not found a real workload where it makes a positive difference.

I think the Tremor guys found it fractionally faster.  @darach @licenser

I think with huge pages it makes more difference particularly for 2^n sized allocations.  We interact much better with those now, so I think it might be better to have this.  The cost of doing it is very small, and the code complexity is not much.

As @davidchisnall says, the meta-data being at the start of the block pushes many things into the same associativity set in the cache.  By using other cache lines in a block it means the `snmalloc` part of the work load is more distributed over the cache. I haven't run tests with this enabled for a while to analyse the results.  

Without huge pages it didn't make much difference as the lower level caches are physically index, so that introduced sufficient noise that the cache effects were minimal.  But with huge pages, the physical indexes are very uniform, and the cache effects could be noticeable.  We did have a long section in the associated paper on this, but had to drop it due to space limits.

**Comment by Licenser** [comment]
We'll revalidate this :) as code changes numbers change we'll give some feedback.

**Comment by darach** [comment]
We found significant benefits with snmalloc for our use cases with tremor compared to mimalloc and jemalloc and we now use snmalloc by default

**Comment by Licenser** [comment]
On our go to (longest kept) benchmark on a Threadriper 24 core we get:

for both

* pinned to core 1-3 (the same CCX shared l3 cache)
* not pinned (24 + 24 threads with infinity fabric for communication)

over the 'huge' sample size of 4 runs for each benchmark run the non `cache-friendly` code was ~1% faster then the `cache-friendly` one.

It is worth mentioning that since we originally adopted snammloc and benchmaked configurations we re-did the entire threading moddle and moved from direct threads to a executor for async tasks. That might be the reason why `cache-friendly` is now slighty slower in this benchmark.

**Comment by mjp41** [comment]
@licenser, thanks super useful to know.

**Comment by photoszzt** [comment]
Thanks for the explanation. Since you mention about huge pages, does snmalloc relies on Transparent huge page? Does it do its own allocation via mmap?

**Comment by mjp41** [comment]
Yes, we rely on Transparent huge pages.  I have not looked into using huge pages directly as mimalloc can do

https://github.com/microsoft/mimalloc/blob/41683071c1395340b4243fb6b4f1133966c13f71/src/os.c#L392-L406

We call mmap here on POSIX platforms:

https://github.com/microsoft/snmalloc/blob/8990c349119bc848fe3c776ce02888d82abdef7f/src/pal/pal_posix.h#L210-L236

This gets overridden in a few POSIX platforms such as BSD platforms that support aligned mmap calls:

https://github.com/microsoft/snmalloc/blob/8990c349119bc848fe3c776ce02888d82abdef7f/src/pal/pal_bsd_aligned.h#L31-L52

This could be overridden on other platforms to use huge pages directly, but we haven't done that.

**Comment by davidchisnall** [comment]
Note that we do try to allocate chunks of virtual address space that are aligned at least on a superpage boundary, so operating systems that do a good job of transparent superpage promotion should get the benefits of both low TLB pressure and not too much internal fragmentation.  Unfortunately, last time I checked, the set of operating systems that did a good job of transparent superpage promotion was just FreeBSD.

**Comment by mjp41** [comment]
Based on the mimalloc-bench switching on transparent huge pages does make things go faster on Linux.
